### ISSUE TITLE
[cmd/mdatagen] Support for parsing and associated error handling

### ISSUE THREAD

**Issue by djaglowski** [open-telemetry/opentelemetry-collector-contrib#7574]
A somewhat common pattern in scrapers involves parsing a numeric value from a string. When doing so, one should check whether the parsing operation succeeded. If it did, then a data point can be recorded. Otherwise, a partial error should be captured.

A very useful construct, `MetricsBuilder`, has recently been added to the `mdatagen` package. This has done a lot to standardize scraper code but we still see a common repetitive pattern when it comes to scrapers that must do a bit more work to obtain `int64` or `float64` values. More or less looks like this:

```go
var errors scrapererror.ScrapeErrors

if f, err := parseFloat(numOne); err != nil {
	errors.AddPartial(1, err)
} else {
	mb.RecordMetricOneDataPoint(now, f)
}

if i, err := parseInt(numTwo); err != nil {
	errors.AddPartial(1, err)
} else {
	mb.RecordMetricTwoDataPoint(now, i)
}
```

There are usually ways to condense this code, but it would be nice if we can provide a concise way to handle this pattern, to further streamline scraper development and encourage commonality across many distinct scrapers.

**Describe the solution you'd like**

I believe the `mdatagen` package can handle this pattern in a reasonable and generic way. Specifically, I suggest that each of the `MetricsBuilder.Record*` functions could have a complementary function that implements a standard parsing pattern. 

More or less, the code would look like: 

```go
func (m MetricsBuilder) ParseMetricOneDataPoint(ts pdata.Timestamp, val string, errors scrapererror.ScrapeErrors) {
	if f, err := parseFloat(val); err != nil {
		errors.AddPartial(1, err)
	} else {
		m.RecordMetricOneDataPoint(now, f)
	}
}

func (m MetricsBuilder) ParseMetricTwoDataPoint(ts pdata.Timestamp, val string, errors scrapererror.ScrapeErrors) {
	if i, err := parseInt(val); err != nil {
		errors.AddPartial(1, err)
	} else {
		m.RecordMetricTwoDataPoint(now, i)
	}
}
```

The implementation of the original example then becomes:
```go
var errors scrapererror.ScrapeErrors
mb.ParseMetricOneDataPoint(now, numOne, errors)
mb.ParseMetricTwoDataPoint(now, numTwo, errors)
```

### Generalized parsing

We could go a step further with this and parse from `interface{}` in a very generic way:

```go
func (m MetricsBuilder) ParseMetricTwoDataPoint(ts pdata.Timestamp, val interface{}) {
	if i, err := parseInt(val); err != nil {
		errors.AddPartial(1, err)
	} else {
		m.RecordMetricTwoDataPoint(now, i)
	}
}

func parseInt(val interface{}) (int64, error) {
	switch val.(type) {
	case string:
		return strconv.ParseInt(value, 10, 64)
	case int:
		return int64(val)
	case int32:
		return int64(val)
	... other data types with reasonable conversion to int64
	default:
		// return error
	}
}
```

### Developer experience considerations

Since many scrapers will not be interested in this functionality, we could attach these functions to a different struct (e.g. `MetricsParser`), which may embed the `MetricsBuilder`. Possibly, the existence of this `MetricsParser` struct can be requested by an additional flag in the go generate tag. e.g. `//go:generate mdatagen --experimental-gen --with-auto-parsers metadata.yaml`

**Comment by djaglowski** [comment]
cc: @dmitryax

**Comment by dmitryax** [comment]
I like the general idea.

Q about Generalized parsing: 

Do we have any scrapers where we don't know a metric value type? It would mean that value is actually coming with `interface{}` type, right? Having this generalized parsing approach introduces some computational overhead, so I'm trying to understand if it's really needed.

**Comment by djaglowski** [comment]
My intention with the generalized parsing suggestion was to reduce complexity for scraper developers, with the idea being that they can throw a variety of types into this function. (i.e. not necessarily that they have an `interface{}` that they want to parse.) I'm not aware of any cases that would require this kind of functionality, so based on your point about the performance overhead, this probably isn't worth doing.

What I really care about here is the we reduce the repetition of string processing.

**Comment by codeboten** [comment]
@djaglowski can this be closed w/ #8986 having merged?
